core/primitives/src/sharding.rs

```diff
@@ -583,15 +583,18 @@ impl ShardChunkHeader {
     }
 
     /// Returns whether the header is valid for given `ProtocolVersion`.
-    pub fn valid_for(&self, version: ProtocolVersion) -> bool {
+    pub fn valid_for(
+        &self,
+        version: ProtocolVersion,
+    ) -> Result<(), BadHeaderForProtocolVersionError> {
```

Comment:
The name doesn't really match the return value. I would suggest either keeping the bool return value or renaming to something like `validate_version`.